### PR TITLE
feat: e2e test setup (cypress / synpress)

### PR DESCRIPTION
**Pull request by domw30**
[Associated Notion Card](https://wilderworld.atlassian.net/jira/software/c/projects/ZAP/boards/34?modal=detail&selectedIssue=ZAP-68)

-----

## 1. Pull request checklist

- [x] Notion card has been moved to the Code Review column
- [x] Notion card has a link to this PR
- [x] A reviewer has been assigned to the Notion card

## 2. PR type
- Feature (testing)
 

## 3. What is the old behaviour?
End to end testing had not been set up / added.

## 4. What is the new behaviour?
- Cypress and Synpress setup for end to end testing.
- Set up custom commands.
- Configured `@testing-library/cypress` to use dom-testing-library helpers for Cypress i.e. `findByText`.
- Added configuration to use `data-test-id` selector in application code for test purposes.
- Updated `README` with testing instructions.
- Initial visit root domain and connect wallet tests added.
- Initial subdomain navigation tests added (WIP)

Screenshot 2023-03-27 at 21 20 43

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Environment Variables used for Cypress e2e testing
+CYPRESS_REMOTE_DEBUGGING_PORT=9222
+NETWORK_NAME=goerli
+SECRET_WORDS='test test test test test test test test test test test junk'
```
Comment (colbr) [inline .env.example:4]: nit: can we change the name to `TEST_WALLET_SEED` or similar? Just to be a bit more clear as to what it is

File: cypress/e2e/connect_wallet.cy.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		cy.get('div').should('have.class', 'wallet-select__header');
+		cy.contains('connect to a wallet', { matchCase: false }).should(
+			'be.visible',
+		);
+		cy.get('.dialog__content')
+			.should('contain', 'Metamask')
+			.and('contain', 'Wallet Connect')
+			.and('contain', 'Coinbase Wallet')
+			.and('contain', 'Fortmatic')
+			.and('contain', 'Portis');
```
Comment (colbr) [inline cypress/e2e/connect_wallet.cy.ts:29]: Is it possible to write these in a way which is more similar to how the user would navigate the website?

i.e. `find a button with the text "Connect To A Wallet", click it, check if there's a modal with the text "Metamask", "Wallet Connect", etc.`

This way we're not tied to implementation details, making tests less brittle.

Similar comments for other tests - would be good if we could use `getByText` as much as possible although I know that can be difficult!

File: cypress/e2e/site_navigation.cy.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import * as domainsPageSelectors from '../../src/pages/Domains/selectors';
+import * as viewSubdomainsSelectors from '../../src/features/view-subdomains/selectors';
+
+describe('Site Navigation', () => {
```
Comment (colbr) [inline cypress/e2e/site_navigation.cy.ts:4]: I wonder if we could use the SDK in these tests to validate results, e.g.

```
subdomains = get subdomains by zna (wilder)
open wilder domain
expect to see the title of each domain
expect to be able to click each domain
```

File: src/features/view-subdomains/selectors.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export const viewSubdomainsSection = 'view-subdomains-section';
+export const domainBannerContainer = 'domain-banner-container';
+export const domainDetailsCardContainer = 'domain-details-card-container';
```
Comment (colbr) [inline src/features/view-subdomains/selectors.ts:3]: nit: I'd recommend something like the below

```ts
enum SubdomainViewSelector = {
    VIEW_SUBDOMAINS_SECTION = 'view-subdomains-section'
    ...
}
```

**Review by colbr — APPROVED** [review]
Looking nice! A couple of nits and questions, but I'm liking the setup.

One thing - it would be really nice if tests could live in the feature folders. That way, feature folders would contain everything needed for the feature including tests. Not really a huge deal, but I think it could be nice for DX!

File: cypress/e2e/site_navigation.cy.ts (hunk 3 above)
Comment (colbr) [reply cypress/e2e/site_navigation.cy.ts:4]: Just remembered this PR is for setup, and not actual tests. Ignore this :)

File: .env.example (hunk 1 above)
Comment (domw30) [reply .env.example:4]: Unfortunately I don't think we can change the variable names here - if changed Synpress doesn't pick get the values. There might be a way to change this in the config but did some digging and haven't been successful

File: cypress/e2e/connect_wallet.cy.ts (hunk 2 above)
Comment (domw30) [reply cypress/e2e/connect_wallet.cy.ts:29]: Yes for sure 🙂 working to improve these now in the follow up PR

File: .env.example (hunk 1 above)
Comment (colbr) [reply .env.example:4]: All good!

**Pull request merged**